File: backend/services/core-api/internal/domain/product.go
package domain

import "time"

// PricingType is how a product's price is determined at sale time.
type PricingType string

const (
	// PricingPerKg products are weighed and priced by a certified scale —
	// UnitPriceCents is the price per kilogram sent to it.
	PricingPerKg PricingType = "per_kg"
	// PricingPerPiece products are sold by count, never touching a scale —
	// UnitPriceCents is the price per single item, and the line's total is
	// ordinary quantity x price arithmetic (no legal-metrology concern,
	// since no physical measurement is involved).
	PricingPerPiece PricingType = "per_piece"
)

// Product is one item a tenant sells, priced either per kilogram (weighed on
// a scale) or per piece (counted).
type Product struct {
	ID             string      `json:"id"`
	TenantID       string      `json:"tenant_id"`
	Name           string      `json:"name"`
	PricingType    PricingType `json:"pricing_type"`
	UnitPriceCents int         `json:"unit_price_cents"` // per kg if PricingType is PricingPerKg, per piece if PricingPerPiece
	CreatedAt      time.Time   `json:"created_at"`
}

File: mobile/lib/models/product.dart
/// Mirrors core-api's domain.Product JSON shape (see
/// backend/services/core-api/internal/domain/product.go).
class Product {
  final String id;
  final String tenantId;
  final String name;

  /// "per_kg" (weighed on a scale) or "per_piece" (counted).
  final String pricingType;

  /// Price per kg if [pricingType] is "per_kg", price per piece if
  /// "per_piece".
  final int unitPriceCents;
  final DateTime createdAt;

  const Product({
    required this.id,
    required this.tenantId,
    required this.name,
    required this.pricingType,
    required this.unitPriceCents,
    required this.createdAt,
  });

  bool get isPerPiece => pricingType == 'per_piece';

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] as String,
      tenantId: json['tenant_id'] as String,
      name: json['name'] as String,
      pricingType: json['pricing_type'] as String,
      unitPriceCents: json['unit_price_cents'] as int,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}
